pennylane/typing.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This file contains different PennyLane types."""
+import contextlib
+
+# pylint: disable=import-outside-toplevel, too-few-public-methods
+import sys
+from typing import Union
+
+import numpy as np
+from autograd.numpy.numpy_boxes import ArrayBox
+
+_TensorLike = Union[int, float, bool, complex, bytes, str, list, tuple, np.ndarray, ArrayBox]
+
+
+class TensorLikeMETA(type):
+    """TensorLike metaclass."""
```

Comment:
Since this is power usage of Python that many developers won't be familiar with, it may be useful adding a link or a description about what these dunder methods work.

We could even potentially just add some comments to the page as to why we are doing things this way, so that developers in a year or two will understand what's going on.

Reply:
Added docstrings explaining what the classes/methods do!